Add Hardware Offload Validation as a test option for flows.

wizhaoredhat opened the pull request:
Hardware Offload Validation uses "perf3", "tcpdump" and "ethtool"
to determine whether a flow has been hardware offloaded properly.
The expectation is that little to no packets are received on the
VF representors for the client and/or server pods.

In order to get the VF representor for the pods involved, the ft-tools
pod is used with the "crictl" application to get the POD ID.

The ft-tools pod was updated to include "crictl". Also the tools
daemonset was updated to mount the host filesystem (similar to
openshift's debug pod e.g. "oc debug node/<name>"). Also fixed the
node selector to point to "sriov", otherwise the ft-tools won't be
created for the nodes.

Some flows aren't expected to be hardware offloaded; thus the "HWOL"
option would not perform the validation for those flows.

wizhaoredhat commented:
@Billy99 @bn222

bn222 commented on hwol-logs/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
+*
+# Except this file
+!.gitignore
```
I prefer a top level gitignore

wizhaoredhat replied:
Done. Created a top level gitignore.

bn222 commented on manifests/tools-daemonSet.yaml.j2:
```diff
@@ -19,11 +19,20 @@ spec:
       containers:
       - name: ft-tools
-        image: nicolaka/netshoot
+        image:  quay.io/wizhao/ft-tools:0.9-x86_64
```
remove 1 space

wizhaoredhat replied:
Done.

bn222 commented on process-cmds.sh:
```diff
@@ -94,6 +94,166 @@ process-iperf() {
+  echo "TX Packets: ${txpktend} - ${txpktstart} = ${txcount}"
+
+  # Wait for Iperf to finish
+  wait $IPERF_PID
```
Why not get ethtool stats after perf finishes?

wizhaoredhat replied:
The TCP FIN/ACK packets aren't offloaded I believe. Just to avoid counting those packets.

Billy99 commented on process-cmds.sh:
```diff
@@ -94,6 +94,166 @@ process-iperf() {
+
+  HWOL_VALIDATION_FILENAME="${HW_OFFLOAD_LOGS_DIR}/${TEST_FILENAME}"
+
+  echo "kubectl get pods -n ${FT_NAMESPACE} --selector=name=${TOOLS_POD_NAME} -o wide | grep -w \"${TEST_CLIENT_NODE}\" | awk -F' ' '{print $1}'"
```
Example of output I would not print, or at least hide behind FT_DEBUG if I was having issues.

wizhaoredhat replied:
Agreed, put it behind FT_DEBUG.

Billy99 commented on process-cmds.sh:
```diff
@@ -94,6 +94,166 @@ process-iperf() {
+  HWOL_VALIDATION_FILENAME="${HW_OFFLOAD_LOGS_DIR}/${TEST_FILENAME}"
+
+  echo "kubectl get pods -n ${FT_NAMESPACE} --selector=name=${TOOLS_POD_NAME} -o wide | grep -w \"${TEST_CLIENT_NODE}\" | awk -F' ' '{print $1}'"
+  TOOLS_CLIENT_POD=$(kubectl get pods -n ${FT_NAMESPACE} --selector=name=${TOOLS_POD_NAME} -o wide | grep -w "${TEST_CLIENT_NODE}" | awk -F' ' '{print $1}')
```
I use a new pattern in [variables.sh-L609](https://github.com/Billy99/ovn-kuber-traffic-flow-tests/blob/main/variables.sh#L609) where I try to limit the kubectl calls to speed it up some. I haven't gone back and fixed other places, but try to follow it in new code. So I pipe the base kubectl command into a TMP_VAR, the run multiple "echo TMP_VAR | grep ... | awk ...". Example here:
```suggestion
  TMP_OUTPUT=$(kubectl get pods -n ${FT_NAMESPACE} --selector=name=${TOOLS_POD_NAME} -o wide)
  TOOLS_CLIENT_POD=$(echo "${TMP_OUTPUT}" | grep -w "${TEST_CLIENT_NODE}" | awk -F' ' '{print $1}')
  TOOLS_SERVER_POD=$(echo "${TMP_OUTPUT}" | grep -w "${TEST_SERVER_NODE}" | awk -F' ' '{print $1}')  
```

Billy99 commented on process-cmds.sh:
```diff
@@ -94,6 +94,166 @@ process-iperf() {
+  return $retVal
+}
+
+process-hw-offload-validation() {
```
Based on the README, there is a lot of output by this command. Once I get everything working, I tend to trim the output back to: 1) commands I need to rerun the test manually 2) SUCCESS or FAILED
For 1), yours is more complicated because of the waiting and reading of counters, so you will need to decide what is important to reproduce command manually, if it is possible. Also, you may not be able to silence some of what the commands spit out, but just looking at the README, there is a lot there. That is what the files are for. Use a combination of FT_DEBUG and VERBOSE to allow more output when needed.

wizhaoredhat replied:
I trimmed the output. I used FT_DEBUG for some of the intermediate commands to determine the VF rep and other variables.

Billy99 commented on README.md:
````diff
@@ -681,6 +690,243 @@ in and set by the script. Example:
+the beginning and end of the `iperf3` duration, `tcpdump` on the VF representor is run
+during the `iperf3` duration, lastly a summary of the results is printed.
+
+```
````
Make sure this is the latest output. I see the following in code but not here:
echo "Client Pod on Client Host VF Representor Results:" > "${HWOL_VALIDATION_FILENAME}"
echo "Client Pod on Server Host VF Representor Results:" >> "${HWOL_VALIDATION_FILENAME}"
echo "Server Pod on Server Host VF Representor Results:" >> "${HWOL_VALIDATION_FILENAME}"

wizhaoredhat replied:
Done

wizhaoredhat commented on manifests/tools-daemonSet.yaml.j2:
```diff
@@ -19,11 +19,20 @@ spec:
       containers:
       - name: ft-tools
-        image: nicolaka/netshoot
+        image:  quay.io/wizhao/ft-tools:0.9-x86_64
```
Done, I added both arm and x86 images to quay.

wizhaoredhat replied:
Also IMAGES.md was updated.

bn222 replied:
if you added both arm and x86, then please change the naming of the image on quay

wizhaoredhat replied:
Yes, it is already fixed if you look at the latest commits.

wizhaoredhat commented on manifests/tools-daemonSet.yaml.j2:
```diff
@@ -19,11 +19,20 @@ spec:
+        name: host
       nodeSelector:
-        ft.ClientPod: client
+        ft.ClientPod: sriov
```
This works, I tried it.

wizhaoredhat commented:
- Some issue with the .gitignore top level file. You cannot have empty folders in git, thus we need subdirectory .gitignore files.
- Adding reverse Iperf tests.

bn222 commented on process-cmds.sh:
```diff
@@ -74,12 +74,15 @@ process-iperf() {
+
+  echo "${MY_CLUSTER}:${TEST_CLIENT_NODE} -(Reverse)-> ${TEST_SERVER_CLUSTER}:${TEST_SERVER_NODE}"
+  echo "kubectl exec -n ${FT_NAMESPACE} ${TEST_CLIENT_POD} -- ${TASKSET_CMD} ${IPERF_CMD} -R -c ${TEST_SERVER_IPERF_DST} -p ${TEST_SERVER_IPERF_DST_PORT} -t ${IPERF_TIME}"
+  kubectl exec -n "${FT_NAMESPACE}" "$TEST_CLIENT_POD" -- /bin/sh -c "${TASKSET_CMD} ${IPERF_CMD} -R -c ${TEST_SERVER_IPERF_DST} -p ${TEST_SERVER_IPERF_DST_PORT} -t ${IPERF_TIME}"  >> "${IPERF_FILENAME}"
```
Below you are using ${IPERF_OPT} and here you are hardcoding "-R". Can we have something more streamlined?

wizhaoredhat replied:
I am not sure what you had in mind. Could you give an example? I will then implement it.

bn222 replied:
I added another comment to point to the inconsistency: You've added -R in IPERF_OPT elsewhere, but here you are doing something else. Why is there a difference?

wizhaoredhat replied:
I amended the last commit. I think I understood what you meant by streamlining. Let me know what you think.

bn222 replied:
looks ok, thanks

wizhaoredhat committed:
Update ft-tools to include "crictl" and to be used for "sriov/client"

The ft-tools pod was updated to include "crictl". Also the tools
daemonset was updated to mount the host filesystem (similar to
openshift's debug pod e.g. "oc debug node/<name>"). Also fixed the
node selector to point to either "sriov" or "client" label, otherwise
the ft-tools won't be created for the nodes. The ft-tools image
on quay was built for both x86_64 and ARM.

Updated the IMAGES.md documentation detailing how to build the
ft-tools image.

wizhaoredhat committed:
Removed "-it" (interactive) from kubectl commands.

The "-it" (interactive) option should not be used in kubectl
commands because none of the commands require user interaction.

wizhaoredhat committed:
Add Hardware Offload Validation and Iperf Rev. Traffic Direction Tests

Hardware Offload Validation uses "perf3", "tcpdump" and "ethtool"
to determine whether a flow has been hardware offloaded properly.
The expectation is that little to no packets are received on the
VF representors for the client and/or server pods.

In order to get the VF representor for the pods involved, the ft-tools
pod is used with the "crictl" application to get the POD ID.

Some flows aren't expected to be hardware offloaded; thus the "HWOL"
option would not perform the validation for those flows.

Iperf reverse traffic direction tests are additional tests performed
along side the forward Iperf traffic direction (Client sends data to
server). This is applicable to both IPERF tests and HWOL tests.

The README.md documentation is updated describing how to use the
Hardware Offload Validation and how reverse Iperf reverse traffic
direction will be used.

wizhaoredhat commented on process-cmds.sh:
```diff
@@ -93,6 +97,208 @@ process-iperf() {
+    echo -e "${GREEN}SUCCESS${NC}\r\n"
+  fi
+
+  IPERF_OPT="-R"
```
Got it.

Billy99 commented on process-cmds.sh:
```diff
@@ -74,12 +74,15 @@ process-iperf() {
+  echo "kubectl exec -n ${FT_NAMESPACE} ${TEST_CLIENT_POD} -- ${TASKSET_CMD} ${IPERF_CMD} ${IPERF_FORWARD_TEST_OPT} -c ${TEST_SERVER_IPERF_DST} -p ${TEST_SERVER_IPERF_DST_PORT} -t ${IPERF_TIME}"
+  kubectl exec -n "${FT_NAMESPACE}" "$TEST_CLIENT_POD" -- /bin/sh -c "${TASKSET_CMD} ${IPERF_CMD} ${IPERF_FORWARD_TEST_OPT} -c ${TEST_SERVER_IPERF_DST} -p ${TEST_SERVER_IPERF_DST_PORT} -t ${IPERF_TIME}"  > "${IPERF_FILENAME}"
+
+  echo "${MY_CLUSTER}:${TEST_CLIENT_NODE} -(Reverse)-> ${TEST_SERVER_CLUSTER}:${TEST_SERVER_NODE}"
```
The check at the end will not verify both tests (client -> server and server -> client) because output is in same file and only grepping for `sender`. I would pipe each test into a separate file (i.e. 01-a-client-server-pod2pod-sameNode.txt and 01-a-server-client-pod2pod-sameNode.txt), and add a check for each.

Billy99 reviewed (commented):
Few nits.

wizhaoredhat committed:
Create seperate log files for different Iperf traffic directions

Each traffic flow direction will log its output in a different file
either with "client-server" or "server-client" in its name.

Refactor output to be better visualized. Also corrected grammer and
typos throughout the files.

wizhaoredhat commented on README.md:
```diff
@@ -307,6 +308,9 @@ This script uses ENV Variables to control test. Here are few key ones:
                                  VERBOSE=true ./test.sh
   IPERF                      - 'iperf3' can be run on each flow, off by default. Example:
                                  IPERF=true ./test.sh
+  HWOL                       - Hardware Offload Validation can be run on each applicable flow."
```
Done.

wizhaoredhat commented on README.md:
```diff
@@ -307,6 +308,9 @@ This script uses ENV Variables to control test. Here are few key ones:
   IPERF                      - 'iperf3' can be run on each flow, off by default. Example:
                                  IPERF=true ./test.sh
+  HWOL                       - Hardware Offload Validation can be run on each applicable flow."
+                               Parameters from IPERF will be used to generate traffic. Example:"
```
Done.

wizhaoredhat commented on README.md:
```diff
@@ -307,6 +308,9 @@ This script uses ENV Variables to control test. Here are few key ones:
                                  IPERF=true ./test.sh
+  HWOL                       - Hardware Offload Validation can be run on each applicable flow."
+                               Parameters from IPERF will be used to generate traffic. Example:"
+                                 HWOL=true ./test.sh"
```
Done.

wizhaoredhat commented on README.md:
```diff
@@ -593,37 +604,65 @@ SUCCESS
 the rough throughput of each flow. When enabled, `iperf3` is run and a
-summary of the results is printed.
+summary of the results is printed. Both forward and reverse directions
+of traffic is tested. Traffic is first sent from client, then traffic
```
Done.

wizhaoredhat commented on docs/IMAGES.md:
```diff
@@ -70,4 +70,58 @@ sudo podman manifest push ft-base-image-0.9-list quay.io/billy99/ft-base-image:0
 
-This is not a private image. Just using https://github.com/nicolaka/netshoot as the tools image.
+The `ft-tools` pod uses a Centos Stream9 image base with root/privileged access to the host.
+Many tools and debug packages pulled in (Please see images/ft-tools/Containerfile for details).
```
Done.

wizhaoredhat commented on README.md:
```diff
@@ -681,6 +714,327 @@ in and set by the script. Example:
+
+*** 1-b: Pod to Pod (Different Node) ***
+
+admin:worker-advnetlab49 -> admin:worker-advnetlab48
```
Done.

wizhaoredhat commented on images/ft-tools/Containerfile:
```diff
@@ -0,0 +1,8 @@
+FROM quay.io/centos/amd64:stream9
```
As discussed, space is a concern. We will monitor and see if we find ourselves having to add more and more tools on top of the existing image.

Billy99 commented:
/lgtm
